diffkemp/output.py

```diff
@@ -0,0 +1,141 @@
+from diffkemp.semdiff.caching import ComparisonGraph
+from diffkemp.utils import get_end_line, EndLineNotFound
+from diffkemp.semdiff.result import Result
+import os
+import yaml
+
+
+class YamlOutput:
+    """Yaml representation of compare result."""
+
+    def __init__(self, snapshot_dir_old, snapshot_dir_new, result):
+        self.old_dir = os.path.join(
+            os.path.abspath(snapshot_dir_old), "")
+        self.new_dir = os.path.join(
+            os.path.abspath(snapshot_dir_new), "")
+        self.result = result
+        self.output = {}
+        # Sets with symbol names
+        self.function_names = set()
+        self.macro_names = set()
+        # Note: type_names contains tuples (type name, parent name)
+        self.type_names = set()
+
+        self._create_output()
+
+    def save(self, output_dir, file_name):
+        with open(os.path.join(output_dir, file_name), "w") as file:
+            yaml.dump(self.output, file, sort_keys=False)
+
+    def _create_output(self):
+        """Creates yaml representation of compare result."""
+        self.output["old-snapshot"] = self.old_dir
+        self.output["new-snapshot"] = self.new_dir
+
+        self._create_results()
+        self._create_definitions()
+
+    def _create_results(self):
+        """Creates call stacks of not-equal functions."""
+        # list of all not-equal functions callstacks
+        results = []
+        for fun_name, fun_result in self.result.inner.items():
+            if fun_result.kind != Result.Kind.NOT_EQUAL:
+                continue
+            self.function_names.add(fun_name)
+            # list of not-equal functions callstacks for a function
+            diffs = []
+            for called_res in sorted(fun_result.inner.values(),
+                                     key=lambda r: r.first.name):
+                # information about a function which is different
+                diff = {}
+                diff["function"] = called_res.first.name
+                # getting callstacks
+                diff["old-callstack"] = called_res.first \
+                    .callstack.to_output_yaml_with_rel_path(self.old_dir) \
+                    if called_res.first.callstack else []
+                diff["new-callstack"] = called_res.second \
+                    .callstack.to_output_yaml_with_rel_path(self.new_dir) \
+                    if called_res.second.callstack else []
+                # updating symbol names
+                if called_res.first.callstack:
+                    function_names, macro_names, type_names = called_res \
+                        .first.callstack.get_symbol_names(fun_name)
+                self.function_names.update(function_names)
```

Comment:
This fails if the condition in the previous `if` statement is False, because no local variable  `function_names` is defined. Was it meant to be indented?